ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -937,6 +890,323 @@ chainSyncClient mkPipelineDecision0 tracer cfg
+                -- 'mostRecentIntersection'.
+                setCandidate CandidateFragment
+                  { candidateChain        =
+                      -- We only request the next block. We'll only request the
+                      -- block after that if this next block is still
+                      -- insufficient. And so on. This doesn't allow
+                      -- pipelining, but that's OK in disaster mode. This way
+                      -- avoids unnecessary work.
+                      maybe (error "impossible!") fst
+                        $ AF.splitAfterPoint theirFrag p
+                  , candidateIsLowDensity = True
+                  }
+                let ld' = LowDensity kis st toApply RequestReady
+                continueWithState ld' go2
+
+      where
+        LowDensity kis st toApply reqStatus = ld
+
+        KnownIntersectionState {
+            theirFrag
+          } = kis
+
+        -- Get the block, advance the state, and try again to forecast.
+        --
+        -- PREREQUISITE: The block is fetched.
+        advanceAndRetryForecast ::
+          RealPoint blk -> [RealPoint blk] -> m (LedgerViewCheck blk)
+        advanceAndRetryForecast rp toApply' = do
+            -- TODO We don't think ChainDB GC will remove them. Explain why
+            -- not.
+            blk <- fromMaybe (error "impossible!") <$> getBlock rp
+            let st' = tickThenReapply lCfg blk st
+            case tryForecastLedgerView (Proxy @blk) lCfg st' hdrSlot of
+              Just v  -> return $ LedgerViewAcquired kis v
+              Nothing -> do
+                  let ld' = LowDensity kis st' toApply' RequestStale
+                  continueWithState ld' go2
+
+    -- Restart the STM transaction.
+    --
+    -- Note that this STM transaction is sensitive simultaneously to both the
+    -- current chain ('stillIntersectsWithCurrentChain') and also which of the
+    -- peer's blocks have been fetched (via 'go1').
+    go2 :: Stateful m (LowDensity blk) (LedgerViewCheck blk)
+    go2 = Stateful $ \ld -> join $ atomically $ do
+        let LowDensity kis _st _toApply _reqStatus = ld
+        stillIntersectsWithCurrentChain cfg cdbView kis >>= \case
+          LostIntersection     -> return $ return NoLongerIntersects
+          NewIntersection kis' ->
+              -- We throw away our advance 'LedgerState' and resume from that
+              -- of the new intersection. We could check if we can carry on
+              -- from where we are, but it doesn't seem worth the extra
+              -- complexity.
+              --
+              -- Note that, if we could have kept the old state, the loop will
+              -- most likely catch up to where it was without having to refetch
+              -- any blocks.
+              continueWithState kis' $
+                checkDensityOrContinue cfg cdbView hdrSlot go1
+          SameIntersection     -> continueWithState ld go1
```

Comment:
We bother with `SameIntersection` to avoid a quadratic complexity, if only on principle (though maybe we're avoiding an attack too).

TODO: document that motivation